ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,439 @@
+  handle).
+
+### Workflows in Client
+
+Early thoughts on workflows are that they'll be Ractors. For the purposes of this phase, all that may matter is how
+workflows are defined so they can be used in a type-safe/custom-named way from a client. All SDKs support defining
+workflows without their body and Ruby should as well.
+
+Most of this TBD until next phase, but for early designs for referencing from clients, this probably looks like:
+
+```ruby
+class MyWorkflow < Temporalio::Workflow
+  workflow_query_attr_reader :some_other_query
+
+  def execute(arg)
+    raise NoMethodError
+  end
+
+  workflow_signal
+  def some_signal(arg)
+    raise NoMethodError
+  end
+
+  workflow_query 'custom-query-name'
+  def some_query(arg)
+    raise NoMethodError
+  end
+end
+```
+
+And client usage may look like:
+
+```ruby
+# Connect client
+client = Client.connect('localhost:7233')
+
+# Start workflow
+handle = client.start_workflow(
+  MyWorkflow,
+  "some arg",
+  id="wf-#{SecureRandom.uuid}",
+  task_queue='my_task_queue',
+)
+
+# Send signal
+handle.signal(MyWorkflow.some_signal, "some arg")
+```
+
+* This leverages a `method_added` approach to "decorate" methods.
+* This will probably create class-method equivalents for each item that will return "definitions" for use in clients.
+* We will not spend much time getting this right at first.
+
+## Data Conversion
+
+Like other SDKs, the following classes will exist:
+
+* `Temporalio::Converter::DataConverter` - Accepts a payload converter (must be Ractor shareable), failure converter
+  (must be Ractor shareable), and a payload codec
+  * 💭 Why use `Temporalio::Converter` module instead of just `Temporalio::DataConverter`? There are many files that are
+    included in conversion and most regular users won't ever use them.
+  * 💭 Why duplicate words instead of `Temporalio::Converter::Data`? `Temporalio::Converter` is not a base class and a
+    `DataConverter` is a completely separate thing from, say, a `PayloadConverter`.
+  * ❓ Would we prefer `Temporalio::Conversion` as the module name?
+* `Temporalio::Converter::PayloadConverter` - Default composite payload implementation accepting a collection of
+  encoding payload converters accepted in constructor, with a default class attr that has a default set.
+  * Users are expected to extend this and override conversion methods as needed. Default instantiation of this is 0
+    encoding converters and a `NotImplementedError` if either is called.
+  * 💭 Why not have a separate base class instead of combining the composite one with the one to override? This was a
+    50/50 call, but most users will probably add an encoding impl instead of override at this level. This is admittedly
+    a slight deviation from other SDKs.
+* `Temporalio::Converter::PayloadConverter::Encoding` - Base unimplemented encoding converter
+  * ❓ Would we prefer `Temporalio::Converter::EncodingPayloadConverter`?
```

Comment:
Don't know what the Ruby standard is here but as you know in our other SDKs we've chosen the longer name.

Reply:
The Ruby standard here is to keep it shorter if already in the module name. Ruby is the only language I can think of that does this commonly. While you can require a file, there is no "importing" a module, it must always be qualified (but people often alias it).